File: app/src/main/java/network/minter/bipwallet/settings/views/SettingsTabPresenter.java
/*
 * Copyright (C) by MinterTeam. 2020
 * @link <a href="https://github.com/MinterTeam">Org Github</a>
 * @link <a href="https://github.com/edwardstock">Maintainer Github</a>
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package network.minter.bipwallet.settings.views;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.view.View;
import android.widget.Switch;

import java.util.concurrent.TimeUnit;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import androidx.biometric.BiometricPrompt;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import moxy.InjectViewState;
import network.minter.bipwallet.R;
import network.minter.bipwallet.analytics.AppEvent;
import network.minter.bipwallet.home.HomeScope;
import network.minter.bipwallet.internal.Wallet;
import network.minter.bipwallet.internal.auth.AuthSession;
import network.minter.bipwallet.internal.dialogs.ConfirmDialog;
import network.minter.bipwallet.internal.dialogs.WalletProgressDialog;
import network.minter.bipwallet.internal.helpers.FingerprintHelper;
import network.minter.bipwallet.internal.helpers.PrefKeys;
import network.minter.bipwallet.internal.mvp.MvpBasePresenter;
import network.minter.bipwallet.internal.settings.SettingsManager;
import network.minter.bipwallet.internal.storage.SecretStorage;
import network.minter.bipwallet.internal.views.list.multirow.MultiRowAdapter;
import network.minter.bipwallet.security.SecurityModule;
import network.minter.bipwallet.settings.contract.SettingsTabView;
import network.minter.bipwallet.settings.repo.MinterBotRepository;
import network.minter.bipwallet.settings.views.rows.SettingsButtonRow;
import network.minter.bipwallet.settings.views.rows.SettingsSwitchRow;
import network.minter.bipwallet.settings.views.rows.TitleRow;
import network.minter.core.MinterSDK;
import timber.log.Timber;

/**
 * minter-android-wallet. 2018
 * @author Eduard Maximovich <edward.vstock@gmail.com>
 */
@HomeScope
@InjectViewState
public class SettingsTabPresenter extends MvpBasePresenter<SettingsTabView> {
    //    private final static int REQUEST_ATTACH_AVATAR = CropImage.CROP_IMAGE_ACTIVITY_REQUEST_CODE;
    private final static int REQUEST_CREATE_PIN_CODE = 1002;
    private final static int REQUEST_VERIFY_PIN_CODE = 1003;
    @Inject AuthSession session;
    @Inject SecretStorage secretStorage;
    @Inject SharedPreferences prefs;
    @Inject SettingsManager settings;
    @Inject FingerprintHelper fingerHelper;
    private MultiRowAdapter mMainAdapter, mAdditionalAdapter, mSecurityAdapter;
    private MinterBotRepository mBotRepo = new MinterBotRepository();

    @Inject
    public SettingsTabPresenter() {
        mSecurityAdapter = new MultiRowAdapter();
        mMainAdapter = new MultiRowAdapter();
        mAdditionalAdapter = new MultiRowAdapter();
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
    }

    @Override
    public void attachView(SettingsTabView view) {
        super.attachView(view);

        getViewState().setSecurityAdapter(mSecurityAdapter);
        getViewState().setMainAdapter(mMainAdapter);
        getViewState().setAdditionalAdapter(mAdditionalAdapter);
        getViewState().setOnOurChannelClickListener(this::onOurChannelClickListener);
        getViewState().setOnSupportChatClickListener(this::onSupportChatClickListener);
    }

    @Override
    public void detachView(SettingsTabView view) {
        super.detachView(view);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_CREATE_PIN_CODE) {
            if (resultCode == Activity.RESULT_OK) {
                Timber.d("PIN successfully handled");
            }
        }
    }

    public void onLogout() {
        getAnalytics().send(AppEvent.SettingsLogoutButton);
        Wallet.app().session().logout();
        Wallet.app().secretStorage().destroy();
        Wallet.app().storage().deleteAll();
        Wallet.app().storageCache().deleteAll();
        Wallet.app().prefs().edit().clear().apply();
        getViewState().startLogin();
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();

        // security row
        SettingsSwitchRow enablePinRow = new SettingsSwitchRow("Unlock with PIN-code", this::isPinCodeEnabled, this::onEnablePinCode);
        SettingsSwitchRow fingerprintRow = new SettingsSwitchRow("Unlock with fingerprint", this::isFingerprintEnabled, this::onEnableFingerprint);
        fingerprintRow.setEnabled(this::isPinCodeEnabled);

        SettingsButtonRow changePinRow = new SettingsButtonRow("Change PIN-code", "", this::onChangePinClick);
        changePinRow.setEnabled(() -> secretStorage.hasPinCode());

        mSecurityAdapter.addRow(new TitleRow("Security"));
        mSecurityAdapter.addRow(enablePinRow);

        if (fingerHelper.isHardwareDetected()) {
            mSecurityAdapter.addRow(fingerprintRow);
        }
        mSecurityAdapter.addRow(changePinRow);


        mMainAdapter.addRow(new TitleRow("Notifications"));
        mMainAdapter.addRow(new SettingsSwitchRow("Enable sounds", this::isSoundsEnabled, this::onSwitchSounds));
        mMainAdapter.addRow(new SettingsSwitchRow("Enable notifications", this::isNotificationsEnabled, this::onSwitchNotifications));

        mAdditionalAdapter.addRow(new TitleRow("Default Wallet"));
        SettingsButtonRow defWalletRow = new SettingsButtonRow(secretStorage.getMainWallet().toShortString(), "", this::onChangeDefWallet);
        mAdditionalAdapter.addRow(defWalletRow);
    }

    private void onChangeDefWallet(View view, View view1, String s) {

    }

    private void onSupportChatClickListener(View view) {
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://t.me/MinterHelp"));
        getViewState().startIntent(intent);
    }

    private void onOurChannelClickListener(View view) {
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://t.me/MinterTeam"));
        getViewState().startIntent(intent);
    }

    private Boolean isNotificationsEnabled() {
        return settings.getBool(SettingsManager.EnableLiveNotifications);
    }

    private Boolean isSoundsEnabled() {
        return prefs.getBoolean(PrefKeys.ENABLE_SOUNDS, false);
    }

    private Boolean isPinCodeEnabled() {
        return prefs.getBoolean(PrefKeys.ENABLE_PIN_CODE, false);
    }

    private Boolean isFingerprintEnabled() {
        return prefs.getBoolean(PrefKeys.ENABLE_FP, false) && prefs.getBoolean(PrefKeys.ENABLE_PIN_CODE, false);
    }

    private void onChangePinClick(View view, View view1, String s) {
        getViewState().startPinCodeManager(REQUEST_CREATE_PIN_CODE, SecurityModule.PinMode.Change);
    }

    private void onEnableFingerprint(View view, Boolean enabled) {
        if (enabled) {
            if (!fingerHelper.hasEnrolledFingerprints()) {
                ((Switch) view).setChecked(false);
                getViewState().startDialog(ctx -> new ConfirmDialog.Builder(ctx, R.string.fingerprint_dialog_enroll_title)
                        .setText(R.string.fingerprint_dialog_enroll_text)
                        .setPositiveAction(R.string.btn_settings, (dialogInterface, integer) -> {
                            dialogInterface.dismiss();
                            getViewState().startFingerprintEnrollment();
                            return null;
                        })
                        .setNegativeAction(R.string.btn_cancel)
                        .create());

                return;
            }

            getViewState().startBiometricPrompt(new BiometricPrompt.AuthenticationCallback() {
                @Override
                public void onAuthenticationError(int errorCode, @NonNull CharSequence errString) {
                    super.onAuthenticationError(errorCode, errString);
                    view.post(() -> {
                        ((Switch) view).setChecked(false);
                    });
                    if (errorCode == BiometricPrompt.ERROR_CANCELED || errorCode == BiometricPrompt.ERROR_USER_CANCELED || errorCode == BiometricPrompt.ERROR_NEGATIVE_BUTTON) {
                        return;
                    }

                    Timber.e("Unable to auth FP: [%d] %s", errorCode, errString);
                }

                @Override
                public void onAuthenticationSucceeded(@NonNull BiometricPrompt.AuthenticationResult result) {
                    super.onAuthenticationSucceeded(result);
                    prefs.edit().putBoolean(PrefKeys.ENABLE_FP, true).apply();
                    Timber.e("Success to auth FP");
                }

                @Override
                public void onAuthenticationFailed() {
                    super.onAuthenticationFailed();
                    Timber.e("FP auth failed");
                }
            });

            return;
        }

        prefs.edit().putBoolean(PrefKeys.ENABLE_FP, false).apply();
    }

    private void onEnablePinCode(View view, Boolean enabled) {
        if (!secretStorage.hasPinCode()) {
            getViewState().startPinCodeManager(REQUEST_CREATE_PIN_CODE, SecurityModule.PinMode.Creation);
        } else {
            getViewState().startPinCodeManager(REQUEST_CREATE_PIN_CODE, SecurityModule.PinMode.Deletion);
        }
    }

    private void onSwitchNotifications(View view, Boolean enabled) {
        settings.putBool(SettingsManager.EnableLiveNotifications, enabled);
    }

    private void onSwitchSounds(View view, Boolean isChecked) {
        prefs.edit().putBoolean(PrefKeys.ENABLE_SOUNDS, isChecked).apply();
        Wallet.app().sounds().play(R.raw.click_pop_zap);
    }

    private void onRequestFreeCoins() {
        getViewState().startDialog(ctx -> {
            unsubscribeOnDestroy(
                    mBotRepo.requestFreeCoins(secretStorage.getAddresses().get(0))
                            .delay(500, TimeUnit.MILLISECONDS)
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe(this::onRequestFreeCoinsSuccess, t -> {
                                Timber.w(t);
                                onRequestFreeCoinsError(t.getMessage());
                            })
            );

            return new WalletProgressDialog.Builder(ctx, "Get free coins")
                    .setText("In progress...")
                    .create();
        });
    }

    private void onRequestFreeCoinsError(String message) {
        getViewState().startDialog(ctx -> new ConfirmDialog.Builder(ctx, "Can't get free coins")
                .setText(String.format("Our apologies, but we can't send to you free coins: %s", message))
                .setPositiveAction("Close")
                .create());
    }

    private void onRequestFreeCoinsSuccess(MinterBotRepository.MinterBotResult result) {
        if (!result.isOk()) {
            onRequestFreeCoinsError(result.getError());
            return;
        }
        getViewState().startDialog(ctx -> new ConfirmDialog.Builder(ctx, "Get free coins")
                .setText(String.format("We sent to you 100 %s", MinterSDK.DEFAULT_COIN))
                .setPositiveAction("OK")
                .create());
    }

}

File: app/src/main/java/network/minter/bipwallet/settings/views/SettingsTabPresenter.kt
/*
 * Copyright (C) by MinterTeam. 2021
 * @link <a href="https://github.com/MinterTeam">Org Github</a>
 * @link <a href="https://github.com/edwardstock">Maintainer Github</a>
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package network.minter.bipwallet.settings.views

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.view.View
import android.widget.Switch
import androidx.biometric.BiometricPrompt
import moxy.InjectViewState
import network.minter.bipwallet.BuildConfig
import network.minter.bipwallet.R
import network.minter.bipwallet.analytics.AppEvent
import network.minter.bipwallet.home.HomeScope
import network.minter.bipwallet.internal.Wallet
import network.minter.bipwallet.internal.auth.AuthSession
import network.minter.bipwallet.internal.dialogs.ConfirmDialog
import network.minter.bipwallet.internal.helpers.FingerprintHelper
import network.minter.bipwallet.internal.mvp.MvpBasePresenter
import network.minter.bipwallet.internal.settings.*
import network.minter.bipwallet.internal.storage.SecretStorage
import network.minter.bipwallet.internal.views.list.multirow.BindingMultiRowAdapter
import network.minter.bipwallet.security.SecurityModule
import network.minter.bipwallet.settings.contract.SettingsTabView
import network.minter.bipwallet.settings.views.rows.SettingsButtonRow
import network.minter.bipwallet.settings.views.rows.SettingsSwitchRow
import network.minter.bipwallet.settings.views.rows.TitleRow
import timber.log.Timber
import java.util.*
import javax.inject.Inject

/**
 * minter-android-wallet. 2018
 * @author Eduard Maximovich <edward.vstock></edward.vstock>@gmail.com>
 */
@HomeScope
@InjectViewState
class SettingsTabPresenter @Inject constructor() : MvpBasePresenter<SettingsTabView>() {
    @Inject lateinit var session: AuthSession
    @Inject lateinit var secretStorage: SecretStorage
    @Inject lateinit var prefs: SharedPreferences
    @Inject lateinit var settings: SettingsManager
    @Inject lateinit var fingerHelper: FingerprintHelper

    private val securityAdapter: BindingMultiRowAdapter = BindingMultiRowAdapter()
    private val mainAdapter: BindingMultiRowAdapter = BindingMultiRowAdapter()
    private val additionalAdapter: BindingMultiRowAdapter = BindingMultiRowAdapter()

    private val isNotificationsEnabled: Boolean
        get() = settings[EnableLiveNotifications]

    private val isSoundsEnabled: Boolean
        get() = settings[EnableSounds]

    private val isPinCodeEnabled: Boolean
        get() = settings[EnablePinCode]

    private val isFingerprintEnabled: Boolean
        get() = settings[EnableFingerprint] && settings[EnablePinCode] && fingerHelper.hasEnrolledFingerprints()

    private val isStoriesEnabled: Boolean
        get() = settings[EnableStories] && BuildConfig.ENABLE_STORIES

    override fun attachView(view: SettingsTabView) {
        super.attachView(view)
        viewState.setSecurityAdapter(securityAdapter)
        viewState.setMainAdapter(mainAdapter)
        viewState.setAdditionalAdapter(additionalAdapter)
        viewState.setOnOurChannelClickListener { onOurChannelClickListener() }
        viewState.setOnSupportChatClickListener { onSupportChatClickListener() }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode != Activity.RESULT_OK) {
            return
        }
        if (requestCode == REQUEST_CREATE_PIN_CODE) {
            Timber.d("PIN successfully settled")
            securityAdapter.clear()
            mainAdapter.clear()
            addRows()
        } else if (requestCode == REQUEST_REMOVE_PIN_CODE) {
            Timber.d("PIN successfully removed")
            settings[EnableFingerprint] = false
            securityAdapter.clear()
            mainAdapter.clear()
            addRows()
        }
    }

    fun onLogout() {
        viewState.startDialog {
            ConfirmDialog.Builder(it, R.string.dialog_title_logout)
                    .setDescription(R.string.dialog_description_logout)
                    .setText(R.string.dialog_text_logout)
                    .setPositiveAction(R.string.btn_logout) { d, _ ->
                        analytics.send(AppEvent.SettingsLogoutButton)
                        Wallet.app().session().logout()
                        Wallet.app().secretStorage().destroy()
                        Wallet.app().storage().deleteAll()
                        Wallet.app().storageCache().deleteAll()
                        Wallet.app().prefs().edit().clear().apply()

                        d.dismiss()
                        viewState.startLogin()

                    }
                    .setNegativeAction(R.string.btn_cancel)
                    .create()
        }

    }

    private fun addRows() {
        // security row
        val enablePinRow = SettingsSwitchRow(
                R.string.settings_sec_pin_enabled_label,
                { isPinCodeEnabled },
                { _, enabled -> onEnablePinCode(enabled) }
        )

        val fingerprintRow = SettingsSwitchRow(
                R.string.settings_sec_fingerprint_enabled_label,
                { isFingerprintEnabled },
                { view, enabled -> onEnableFingerprint(view, enabled) }
        )

        fingerprintRow.setEnabled { isPinCodeEnabled }

        val changePinRow = SettingsButtonRow(
                R.string.settings_sec_pin_change_label,
                ""
        ) { view, sharedView, value ->
            onChangePinClick(view, sharedView, value)
        }

        securityAdapter.add(TitleRow(R.string.settings_title_security))
        securityAdapter.add(enablePinRow)

        if (secretStorage.hasPinCode()) {
            if (fingerHelper.isHardwareDetected) {
                securityAdapter.add(fingerprintRow)
            }
            securityAdapter.add(changePinRow)
        }

        mainAdapter.add(TitleRow(R.string.settings_title_notifications))

        mainAdapter.add(SettingsSwitchRow(
                R.string.settings_sounds_enabled_label,
                { isSoundsEnabled },
                { _, isChecked -> onSwitchSounds(isChecked) }
        ))
        mainAdapter.add(SettingsSwitchRow(
                R.string.settings_notifications_enabled_label,
                { isNotificationsEnabled },
                { _: View, enabled: Boolean -> onSwitchNotifications(enabled) }
        ))

        if (BuildConfig.ENABLE_STORIES) {
            mainAdapter.add(TitleRow(R.string.settings_title_stories))
            mainAdapter.add(SettingsSwitchRow(
                    R.string.settings_stories_enabled_label,
                    { isStoriesEnabled },
                    { _, enabled -> onSwitchStories(enabled) }
            ))
        }
    }

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()

        // unset if user disabled fingerprint by OS
        if (isFingerprintEnabled) {
            if (!fingerHelper.hasEnrolledFingerprints() || !fingerHelper.isHardwareDetected) {
                settings[EnableFingerprint] = false
            }
        }

        addRows()

    }

    @Suppress("UNUSED_PARAMETER")
    private fun onChangeDefWallet(value: String) {
    }

    private fun onSupportChatClickListener() {
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse("https://t.me/MinterHelp"))
        viewState.startIntent(intent)
    }

    private fun onOurChannelClickListener() {
        if (Locale.getDefault().isO3Language.lowercase(Locale.getDefault()) == "rus") {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse("https://t.me/MinterNetwork"))
            viewState.startIntent(intent)
        } else {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse("https://t.me/MinterTeam"))
            viewState.startIntent(intent)
        }
    }

    @Suppress("UNUSED_PARAMETER")
    private fun onChangePinClick(view: View, view1: View, s: String) {
        viewState.startPinCodeManager(REQUEST_CREATE_PIN_CODE, SecurityModule.PinMode.Change)
    }

    private fun onEnableFingerprint(view: View, enabled: Boolean) {
        if (enabled) {
            if (!fingerHelper.hasEnrolledFingerprints()) {
                (view as Switch).isChecked = false

                viewState.startDialog { ctx: Context ->
                    ConfirmDialog.Builder(ctx, R.string.fingerprint_dialog_enroll_title)
                            .setText(R.string.fingerprint_dialog_enroll_text)
                            .setPositiveAction(R.string.btn_settings) { d, _: Int ->
                                d.dismiss()
                                viewState.startFingerprintEnrollment()
                            }
                            .setNegativeAction(R.string.btn_cancel)
                            .create()
                }

                return
            }
            viewState.startBiometricPrompt(object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                    super.onAuthenticationError(errorCode, errString)
                    view.post { (view as Switch).isChecked = false }
                    if (errorCode == BiometricPrompt.ERROR_CANCELED || errorCode == BiometricPrompt.ERROR_USER_CANCELED || errorCode == BiometricPrompt.ERROR_NEGATIVE_BUTTON) {
                        return
                    }
                    Timber.e("Unable to auth FP: [%d] %s", errorCode, errString)
                }

                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    super.onAuthenticationSucceeded(result)
                    settings[EnableFingerprint] = true
                    Timber.e("Success to auth FP")
                }

                override fun onAuthenticationFailed() {
                    super.onAuthenticationFailed()
                    Timber.e("FP auth failed")
                }
            })
            return
        }
        settings[EnableFingerprint] = false
    }

    @Suppress("UNUSED_PARAMETER")
    private fun onEnablePinCode(enabled: Boolean) {
        if (!secretStorage.hasPinCode()) {
            if (enabled) {
                viewState.startPinCodeManager(REQUEST_CREATE_PIN_CODE, SecurityModule.PinMode.Creation)
            }

        } else {
            if (!enabled) {
                viewState.startPinCodeManager(REQUEST_REMOVE_PIN_CODE, SecurityModule.PinMode.Deletion)
            }
        }
    }

    private fun onSwitchNotifications(enabled: Boolean) {
        settings[EnableLiveNotifications] = enabled
        Wallet.app().sounds().play(R.raw.click_pop_zap)
    }

    private fun onSwitchStories(enabled: Boolean) {
        settings[EnableStories] = enabled
        Wallet.app().sounds().play(R.raw.click_pop_zap)
        viewState.notifyStoriesState(enabled)
    }

    private fun onSwitchSounds(isChecked: Boolean) {
        settings[EnableSounds] = isChecked
        if (isChecked) {
            Wallet.app().sounds().playForce(R.raw.click_pop_zap)
        }
    }

    companion object {
        private const val REQUEST_CREATE_PIN_CODE = 1002
        private const val REQUEST_REMOVE_PIN_CODE = 1004
        private const val REQUEST_VERIFY_PIN_CODE = 1003
    }
}